vissl/data/airstore_dataset.py

```diff
@@ -0,0 +1,130 @@
+        self.airstore_uri = path
+        self.split = split
+        self.epoch = 0
+        self.start_iter = 0
+        self.enable_queue_dataset = cfg["DATA"][self.split]["ENABLE_QUEUE_DATASET"]
+        self.global_rank = get_rank()
+        self.global_world_size = get_world_size()
+        self._iterator = None
+
+
+    def set_epoch(self, epoch):
+        # set by trainer when train on new epoch or restore from a checkpoint
+        logging.info(f"set epoch to {epoch} in airstore dataset")
+        self.epoch = epoch;
+
+    def set_start_iter(self, start_iter):
+        # set by trainer when train on restoring from a checkpoint
+        self.start_iter = start_iter
+
+    def _open_iterator(self):
+        # data iterator from airstore for current data split.
+        # data are sharded by global total number of workers after shuffling
+        
+        data_cfg = self.cfg["DATA"]
+        split_cfg = data_cfg[self.split]
+
+        worker_info = torch.utils.data.get_worker_info()
+        if worker_info is None:
+            num_workers = 1
+            worker_id = 0
+        else:
+            num_workers = worker_info.num_workers;
+            worker_id = worker_info.id
+
+        return self.pathmgr.opent(
+            self.airstore_uri,
+            "r",
+            skip_samples=self.start_iter * self.batch_size,
+            enable_shuffle=getattr(split_cfg, "AIRSTORE_ENABLE_SHUFFLE", True),
+            shuffle_window=getattr(split_cfg, "AIRSTORE_SHUFFLE_WINDOW", 128),
+            seed=self.epoch,
+            world_size=self.global_world_size * num_workers, # split the dataset for each worker
+            rank=self.global_rank * num_workers + worker_id, # each worker take it's split by it's parent process rank and worker id
+            limit=getattr(split_cfg, "DATA_LIMIT", -1),
+            offset=getattr(split_cfg, "DATA_OFFSET", 0),
+            num_of_threads=getattr(split_cfg, "AIRSTORE_NUM_OF_THREADS", 2),
+            prefetch=getattr(split_cfg, "AIRSTORE_PREFETCH", 1),
+            max_holding_bundles=getattr(split_cfg, "AIRSTORE_MAX_HOLDING_BUNDLES", 5),
+            bundle_download_timeout_ms=getattr(split_cfg, "AIRSTORE_BUNDLE_DOWNLOAD_TIMEOUT_MS", 30000),
+            max_retries=getattr(split_cfg, "AIRSTORE_MAX_RETRIES", 5),
+            dataset_catalog_path=getattr(split_cfg, "AIRSTORE_DS_CATALOG_PATH", None), # temporary need during airstore development
+            env="OSS",
+        )
+
+
+    def num_samples(self):
+        return self._open_iterator().total_size
+
+
+    def __len__(self):
+        return self.num_samples();
+
+    def __getitem__(self, index):
+        if self._iterator is None:
+            self._iterator = self._open_iterator()
+
+        if not self.queue_init and self.enable_queue_dataset:
+            self._init_queues()
+
+        try:
+            # TODO (wpc, prigoyal): we should check images are good when we are
```

Comment:
it would be great to verify this now. Essentially, when we get images from fairstore, we observed the images could be truncated . Correctness of images, if possible, could be tested during the ingestion time in airstore. Thoughts? :)

Reply:
I totally agree this should be handled at the ingestion time. let me talk with the team